packages/go_router/lib/src/information_provider.dart

```diff
@@ -6,6 +6,40 @@ import 'package:flutter/services.dart';
 import 'package:flutter/widgets.dart';
 import 'parser.dart';
 
+/// A data class that can be serialized and stored into browser history entry.
+///
+/// Passing an instance of this class into the `extra` parameter makes the
+/// browser store to memorize data in this instance when handling browser
+/// backward and forward button, e.g pressing backward button in a browser will
+/// send back the BrowserState that associated with the URL.
+///
+/// {@tool snippet}
+/// To create a [BrowserState] with the data, the data must be encoded through
+/// JSON codec.
+///
+/// ```dart
+/// ElevatedButton(
+///   child: const Text('button'),
+///   onPressed: () {
+///     final Map<String, String> data = <String, String>{
+///       'id': '123',
+///     };
+///     final BrowserState extra = BrowserState(
+///       jsonString: const JsonEncoder.convert(data),
+///     );
+///     context.go('/user', extra);
+///   }
+/// )
+/// {@end-tool}
+/// ```
+class BrowserState {
+  /// Creates an [BrowserState] with a JSON encoded string.
+  const BrowserState({required this.jsonString});
+
+  /// A JSON encoded string for the data.
+  final String jsonString;
```

Comment:
What's the value of "restricting" this to JSON? The user could use a non-JSON value, right? Perhaps we should generalize this property's name? For example, the family sample could just `toString` the ID instead of JSON encoding it.

Reply:
yes, any string should work.